jbmc/src/java_bytecode/java_bytecode_parser.cpp

```diff
@@ -562,59 +562,57 @@ void java_bytecode_parsert::get_class_refs()
   for(const auto &field : parse_tree.parsed_class.fields)
   {
     get_annotation_class_refs(field.annotations);
-    typet field_type;
+
     if(field.signature.has_value())
     {
-      field_type=java_type_from_string_with_exception(
+      typet field_type = *java_type_from_string_with_exception(
         field.descriptor,
         field.signature,
-        "java::"+id2string(parse_tree.parsed_class.name));
+        "java::" + id2string(parse_tree.parsed_class.name));
 
       // add generic type args to class refs as dependencies, same below for
       // method types and entries from the local variable type table
       get_dependencies_from_generic_parameters(
         field_type, parse_tree.class_refs);
+      get_class_refs_rec(field_type);
     }
-    else
-      field_type=java_type_from_string(field.descriptor);
-
-    get_class_refs_rec(field_type);
+    else if(const auto field_type = java_type_from_string(field.descriptor))
+      get_class_refs_rec(*field_type);
   }
 
   for(const auto &method : parse_tree.parsed_class.methods)
   {
     get_annotation_class_refs(method.annotations);
     for(const auto &parameter_annotations : method.parameter_annotations)
       get_annotation_class_refs(parameter_annotations);
-    typet method_type;
+
     if(method.signature.has_value())
     {
-      method_type=java_type_from_string_with_exception(
+      typet method_type = *java_type_from_string_with_exception(
         method.descriptor,
         method.signature,
-        "java::"+id2string(parse_tree.parsed_class.name));
+        "java::" + id2string(parse_tree.parsed_class.name));
       get_dependencies_from_generic_parameters(
         method_type, parse_tree.class_refs);
+      get_class_refs_rec(method_type);
     }
-    else
-      method_type=java_type_from_string(method.descriptor);
+    else if(const auto method_type = java_type_from_string(method.descriptor))
+      get_class_refs_rec(*method_type);
```

Comment:
:bulb: Might be worth sticking a warning here that it somehow failed to get the type from both the signature and the descriptor. I'd even be tempted to put an `UNREACHABLE;` here, but I suppose this is one of the cases you were referring to where this broke tests.

Reply:
Indeed, this is reachable/reached in several regression tests. I don't feel sufficiently competent on this code to assess whether that's actually a problem. Any volunteers?

Reply:
I put `UNREACHABLE;` here and didn't get any failures in `jbmc*` - have a specific example? (will try the other two places too).

Reply:
Not sure what exactly your change was, but I'd just use `else { typet method_type = *java_type_from_string(method.descriptor); get_class_refs_rec(method_type); }` here.